examples/3_Advanced_Topics/3-2_Creating_an_XML_BoM.py

```diff
@@ -0,0 +1,247 @@
+﻿# ---
+# jupyter:
+#   jupytext:
+#     formats: ipynb,py:light
+#     text_representation:
+#       extension: .py
+#       format_name: light
+#       format_version: '1.5'
+#       jupytext_version: 1.15.1
+#   kernelspec:
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# # Create an XML BoM using python
+
+# The `bom_types` namespace can be used in conjunction with the `BoMHandler` class to
+# create and manipulate BoMs for analysis. This example shows how this can be done, creating
+# a BoM for a laminated glass door and running a compliance query.
+
+# Most installations of Granta MI will use the default database key and table names:
+
+# +
+DB_KEY = "MI_Restricted_Substances"
+TABLE_NAME = "MaterialUniverse"
+# -
+
+# The structure of an XML BoM is hierarchical, individual parts belong to assemblies which can
+# belong to larger assemblies. It is possible to construct the BoM in one object, but it is often
+# simpler to build the part up from smaller assemblies. This example uses this approach.
+
+# It is possible to have the same part in multiple places in the BoM, so we define two helper functions
+# to create a copy of a part and set the quantity.
+
+# +
+import copy
+
+import ansys.grantami.bomanalytics.bom_types as bom_types
+
+def add_part_to_assembly_with_count(child: bom_types.Part, count: int) -> bom_types.Part:
+    return add_part_to_assembly_with_quantity(child, float(count), "Each")
+
+def add_part_to_assembly_with_quantity(child: bom_types.Part, quantity: float, unit: str) -> bom_types.Part:
+    assigned_part = copy.deepcopy(child)
+    assigned_part.quantity = bom_types.UnittedValue(quantity, unit=unit)
+    return assigned_part
+# -
+
+# Material references are reused, we can define them in different ways.
+# It's common to specify materials by their Granta MI Record GUIDs, which uniquely identify
+# a record.
+# Some databases also have unique identifiers for materials. If these are Short Text attributes
+# they can be used as lookup values, in MaterialUniverse we can use the "Material ID" attribute.
+
+# +
+material_id_reference = bom_types.MIAttributeReference(db_key=DB_KEY, attribute_name="Material ID", table_reference=bom_types.PartialTableReference(table_name=TABLE_NAME))
+
+laminated_glass_reference = bom_types.MIRecordReference(db_key=DB_KEY, record_guid="85ed8b21-c2e6-4c43-8ec3-4c12a44c820c")
```

Comment:
It might be worth splitting these into different cells, with a clarifying sentence above each one. For example:

"The following two references are created using record history GUIDs. The record history GUID identifies the latest accessible version of a record in a Granta MI database".